jfrog-cli/artifactory/utils/golang/project/dependencies/packageswithdeps.go

```diff
@@ -0,0 +1,256 @@
+package dependencies
+
+import (
+	"fmt"
+	golangutil "github.com/jfrog/jfrog-cli-go/jfrog-cli/artifactory/utils/golang"
+	"github.com/jfrog/jfrog-cli-go/jfrog-cli/utils/config"
+	"github.com/jfrog/jfrog-client-go/httpclient"
+	"github.com/jfrog/jfrog-client-go/utils/errorutils"
+	"github.com/jfrog/jfrog-client-go/utils/io/fileutils"
+	"github.com/jfrog/jfrog-client-go/utils/log"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+// Represent go dependency package when running with deps tidy set to true.
+type PackageWithDeps struct {
+	Dependency             *Package
+	transitiveDependencies []PackageWithDeps
+	regExp                 *RegExp
+	runGoModCommand        bool
+	tempDir                string
+	cachePath              string
+}
+
+// Creates a new dependency
+func (pwd *PackageWithDeps) New(cachePath string, dependency Package) GoPackage {
+	pwd.Dependency = &dependency
+	pwd.cachePath = cachePath
+	pwd.transitiveDependencies = nil
+	return pwd
+}
+
+// Performs population of the mod file and publish the dependency and it's transitive dependencies to Artifactory
+func (pwd *PackageWithDeps) PopulateModIfNeededAndPublish(targetRepo string, cache *golangutil.DynamicCache, details *config.ArtifactoryDetails) error {
+	var path string
+	var err error
+	log.Debug("Starting to work on", pwd.Dependency.GetId())
+	dependenciesMap := cache.GetGlobalMap()
+	published, _ := dependenciesMap[pwd.Dependency.GetId()]
+	if published {
+		log.Debug("Overwriting the mod file in the cache from the one from Artifactory", pwd.Dependency.GetId())
+		moduleAndVersion := strings.Split(pwd.Dependency.GetId(), ":")
+		path = downloadModFileFromArtifactoryToLocalCache(pwd.cachePath, targetRepo, moduleAndVersion[0], moduleAndVersion[1], details, httpclient.NewDefaultHttpClient())
+		err = pwd.updateModContent(path, cache)
+		logErrorIfOccurred(err)
+		pwd.runGoModCommand = false
+	} else {
+		pwd.runGoModCommand = !pwd.patternMatched(pwd.regExp.GetNotEmptyModRegex())
+	}
+
+	path, output, err := pwd.createDependencyAndRunGo()
+	logErrorIfOccurred(err)
+	return pwd.publishDependencyAndPopulateTransitive(path, targetRepo, output, cache, details)
+}
+
+// Updating the new mod content
+func (pwd *PackageWithDeps) updateModContent(path string, cache *golangutil.DynamicCache) error {
+	if path != "" {
+		modContent, err := ioutil.ReadFile(path)
+		if err != nil {
+			cache.IncreaseFailures()
+			return errorutils.CheckError(err)
+		}
+		pwd.Dependency.SetModContent(modContent)
+	}
+	return nil
+}
+
+// Init the dependency information if needed.
+func (pwd *PackageWithDeps) Init() error {
+	var err error
+	pwd.regExp, err = GetRegex()
+	if err != nil {
+		return err
+	}
+
+	if !fileutils.IsTempDirInit() {
```

Comment:
The *fileutils.IsTempDirInit()* API can be removed.
The *fileutils.CreateTempDirPath()* should be moved to be earlier in the code flow, followed by the *defer fileutils.RemoveTempDirPath()* API call.

Reply:
Done